File: src/providers/supabase/rules/idempotent-mutations.ts
/**
 * supabase-idempotent-mutations (reliability)
 *
 * `.insert(...)` with no idempotency/dedupe key field is not safely
 * retryable: a flaky network retry, double-click, or browser replay creates
 * a duplicate row. `.upsert(..., { onConflict: ... })` is the documented
 * fix and is exempt — only plain `.insert()` calls are checked.
 */
import { chainObjectCall, memberPropName } from '../utils.js';

function objectHasIdempotencyKey(objectExpression: any): boolean {
  if (objectExpression?.type !== 'ObjectExpression') return false;
  return (objectExpression.properties ?? []).some((p: any) => {
    if (p?.type !== 'Property') return false;
    const name =
      p.key?.type === 'Identifier' ? p.key.name : p.key?.type === 'Literal' ? p.key.value : undefined;
    return typeof name === 'string' && /idempot|dedupe/i.test(name);
  });
}

function insertPayloadHasIdempotencyKey(arg: any): boolean {
  if (arg?.type === 'ObjectExpression') return objectHasIdempotencyKey(arg);
  if (arg?.type === 'ArrayExpression') {
    return (arg.elements ?? []).some((el: any) => objectHasIdempotencyKey(el));
  }
  return false;
}

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Supabase insert calls should be retry-safe via an idempotency key',
      category: 'reliability',
      rationale:
        'Nothing prevents a duplicate row if the client fetch behind an insert is retried (flaky network, double-click, browser replay) — there is no unique constraint or dedupe key visible in the payload, and no upsert semantics. Generate a client-side idempotency key per logical action and either include it as a field guarded by a unique constraint, or use .upsert(..., { onConflict: "<key column>" }).',
      docsUrl: 'https://supabase.com/docs/reference/javascript/upsert',
      recommended: true,
    },
    messages: {
      missingIdempotencyKey:
        'This insert has no idempotency/dedupe key field, so a retried request can create a duplicate row. Add one, or use .upsert(..., { onConflict: "<key column>" }).',
    },
    schema: [],
  },
  create(context: any) {
    return {
      CallExpression(node: any) {
        const prop = memberPropName(node);
        if (prop !== 'insert') return;

        const objCall = chainObjectCall(node);
        if (!objCall || memberPropName(objCall) !== 'from') return;

        const arg = node.arguments?.[0];
        if (!arg) return;
        if (insertPayloadHasIdempotencyKey(arg)) return;

        context.report({ node, messageId: 'missingIdempotencyKey' });
      },
    };
  },
};

export const supabaseIdempotentMutationsRule = rule;
export default rule;

File: src/providers/supabase/rules/python/idempotent_mutations.py
"""supabase-idempotent-mutations

Parity with the JS rule: `.insert(payload)` with no idempotency/dedupe key
field is not safely retryable. `.upsert(..., on_conflict=...)` is the
documented fix and is exempt — only plain `.insert()` calls are checked.
"""
from __future__ import annotations

import ast
import re
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import call_report_loc, chain_object_call, is_supabase_table_call, member_prop_name

RULE_KEY = "supabase-idempotent-mutations"

_KEY_RE = re.compile(r"idempot|dedupe", re.I)
_ID_RE = re.compile(r"^(id|uuid)$", re.I)
_SUFFIX_RE = re.compile(r"_(key|uuid)$", re.I)


def _looks_like_idempotency_key(name: str) -> bool:
    return bool(_KEY_RE.search(name) or _ID_RE.match(name) or _SUFFIX_RE.search(name))


def _dict_has_idempotency_key(d: ast.AST | None) -> bool:
    if not isinstance(d, ast.Dict):
        return False
    for k in d.keys:
        if isinstance(k, ast.Constant) and isinstance(k.value, str) and _looks_like_idempotency_key(k.value):
            return True
    return False


def _payload_has_idempotency_key(arg: ast.AST | None) -> bool:
    if isinstance(arg, ast.Dict):
        return _dict_has_idempotency_key(arg)
    if isinstance(arg, ast.List):
        return any(_dict_has_idempotency_key(el) for el in arg.elts)
    return False


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call) or member_prop_name(node) != "insert":
            continue

        obj_call = chain_object_call(node)
        if not obj_call or not is_supabase_table_call(obj_call):
            continue

        if not node.args:
            continue
        arg = node.args[0]
        if _payload_has_idempotency_key(arg):
            continue

        line, col = call_report_loc(node)
        out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
